issue: Error thrown while handling null values in PPDDM Agent

# Issue report

**Expected Behavior**
Null values should be handled without any error

**Actual Behavior**
Following error has been thrown.

**Steps to Reproduce the Problem**

  1. Try creating model for Use Case 2 by using SAS agent which includes some variables with some null values.

**Logs**
Caused by: org.apache.spark.SparkException: Failed to execute user defined function(VectorAssembler$$Lambda$4824/0x000000084196c840: (                                                                                 struct<Age:double,Gender_VEC:struct<type:tinyint,size:int,indices:array<int>,values:array<double>>,Smokingstatus:double,Heartfailure:d                                                                                 ouble,Hypertension:double,Atrialfibrillation:double,Pulmonaryhypertension:double,Chronickidneydisease:double,Prostatichypertrophy:doub                                                                                 le,Renalcalculi:double,Asthma:double,Obstructivesleepapneasyndrome:double,Pulmonaryembolism:double,Inflammatoryboweldisease:double,Chr                                                                                 onicliverdisease:double,Chronicpancreatitis:double,Gastrooesophagealreflux:double,Diverticulardiseaseofcolon:double,Diabetesmellitus:d                                                                                 ouble,Hyperlipidemia:double,Obesity:double,Hypothyroidism:double,Hyperthyroidism:double,Activeneoplasia:double,HIVAIDS:double,Chronica                                                                                         at scala.util.Success.map(Try.scala:213)
        at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
        at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
        at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
        at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
        at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)

--- a/ppddm-agent/src/main/scala/ppddm/agent/controller/dm/DataAnalysisManager.scala
+++ b/ppddm-agent/src/main/scala/ppddm/agent/controller/dm/DataAnalysisManager.scala
@@ -5,9 +5,9 @@ import org.apache.spark.ml.PipelineStage
 import org.apache.spark.ml.feature.{MinMaxScaler, OneHotEncoder, StringIndexer, VectorAssembler}
 import org.apache.spark.sql.DataFrame
 import org.apache.spark.sql.types.StringType
-import ppddm.core.ai.transformer.{AgeTransformer, MultipleColumnOneHotEncoder}
 import ppddm.agent.controller.prepare.DataPreparationController
-import ppddm.core.rest.model.{DataMiningException, VariableDataType, VariableType}
+import ppddm.core.ai.transformer.{AgeTransformer, MultipleColumnOneHotEncoder}
+import ppddm.core.rest.model.{Algorithm, AlgorithmParameterName, DataMiningException, VariableDataType, VariableType}
 
 import scala.collection.mutable.ListBuffer
 import scala.concurrent.ExecutionContext.Implicits.global
@@ -24,9 +24,10 @@ object DataAnalysisManager {
    *
    * @param dataset_id The id of the Dataset prepared and stored in the data store previously
    * @param dataFrame The original DataFrame retrieved from DataStore
+   * @param algorithm The algorithm
    * @return Array of "PipelineStage"s of several Feature Transformers like StringIndexer, OneHotEncoder, Vector Assembler which create "features" and "label" columns for machine learning algorithms
    */
-  def performDataAnalysis(dataset_id: String, dataFrame: DataFrame): Future[Array[PipelineStage]] = {
+  def performDataAnalysis(dataset_id: String, dataFrame: DataFrame, algorithm: Algorithm): Future[Array[PipelineStage]] = {
     Future {
       logger.debug("Performing data analysis...")
 
@@ -51,6 +52,9 @@ object DataAnalysisManager {
 
       // TODO consider dropping columns in which all the records have the same value
 
+      // options are "keep", "error" or "skip". "keep" puts unseen labels in a special additional bucket, at index numLabels
+      val handleInvalid = algorithm.parameters.find(_.name == AlgorithmParameterName.HANDLE_INVALID).map(_.value).getOrElse("keep")
+
       /**
        * Handle categorical variables
        */
@@ -67,7 +71,7 @@ object DataAnalysisManager {
           .setInputCol(v.variable.getMLValidName)
           .setOutputCol(s"${v.variable.getMLValidName}_INDEX")
           .setStringOrderType("alphabetAsc") // options are "frequencyDesc", "frequencyAsc", "alphabetDesc", "alphabetAsc"
-          .setHandleInvalid("keep") // options are "keep", "error" or "skip". "keep" puts unseen labels in a special additional bucket, at index numLabels
+          .setHandleInvalid(handleInvalid)
       })
       stringIndexerSeq.foreach(i => pipelineStages += i) // Now, DataFrame contains new columns with "_INDEX" at the end of column name
 
@@ -94,6 +98,7 @@ object DataAnalysisManager {
       val vectorAssembler = new VectorAssembler()
         .setInputCols(independentVariables.map(iv => if (categoricalColumns.contains(iv.variable.getMLValidName)) s"${iv.variable.getMLValidName}_VEC" else iv.variable.getMLValidName).toArray) // columns that need to added to feature column
         .setOutputCol("nonScaledFeatures")
+        .setHandleInvalid(handleInvalid)
       pipelineStages += vectorAssembler
 
       // Introduce the dependent variable
@@ -101,7 +106,7 @@ object DataAnalysisManager {
         val labelStringIndexer = new StringIndexer()
           .setInputCol(dependentVariableOption.get.variable.getMLValidName)
           .setOutputCol("label")
-          .setHandleInvalid("keep") // options are "keep", "error" or "skip". "keep" puts unseen labels in a special additional bucket, at index numLabels
+          .setHandleInvalid(handleInvalid)
         pipelineStages += labelStringIndexer
       }
 

--- a/ppddm-agent/src/main/scala/ppddm/agent/controller/dm/algorithm/classification/ClassificationAlgorithm.scala
+++ b/ppddm-agent/src/main/scala/ppddm/agent/controller/dm/algorithm/classification/ClassificationAlgorithm.scala
@@ -35,7 +35,7 @@ trait ClassificationAlgorithm extends DataMiningAlgorithm {
 
     // Prepare the data for execution of the data mining algorithms,
     // i.e. perform the exploratory data analysis which include categorical variable handling, null values handling etc.
-    DataAnalysisManager.performDataAnalysis(dataset_id, dataFrame) map { pipelineStages =>
+    DataAnalysisManager.performDataAnalysis(dataset_id, dataFrame, algorithm) map { pipelineStages =>
 
       // TODO we can also perform Train-Validation Split here if we parameters as array from the client.
 

--- a/ppddm-core/src/main/scala/ppddm/core/rest/model/RestEnum.scala
+++ b/ppddm-core/src/main/scala/ppddm/core/rest/model/RestEnum.scala
@@ -123,6 +123,8 @@ object AlgorithmParameterName extends Enumeration {
   val NUMBER_OF_FOLDS = "num_folds" // Value of k in k-fold Cross validation
   val MAX_PARALLELISM = "max_parallelism" // The maximum level of parallelism to evaluate models in parallel. 1 means serial evaluation
   val METRIC = "metric" // Metric to use on Cross validation
+  val HANDLE_INVALID = "keep" // Options are "keep", "error" or "skip". "keep" puts unseen labels in a special additional bucket, at index numLabels.
+                              // "error" throws an error when there is a null value. "skip" removes the row directly.
 
   val THRESHOLD = "threshold" // Threshold (Double) [Logistic Regression]
   val MAX_ITER = "max_iter" // Maximum number of iterations (Integer) [Logistic Regression, GBT, SVM]

--- a/ppddm-manager/src/main/scala/ppddm/manager/controller/algorithm/AlgorithmController.scala
+++ b/ppddm-manager/src/main/scala/ppddm/manager/controller/algorithm/AlgorithmController.scala
@@ -28,11 +28,13 @@ object AlgorithmController {
     val impurity = Parameter(AlgorithmParameterName.IMPURITY, DataType.STRING, "gini")
     val num_trees = Parameter(AlgorithmParameterName.NUM_TREES, DataType.INTEGER, "5")
     val feature_subset_strategy = Parameter(AlgorithmParameterName.FEATURE_SUBSET_STRATEGY, DataType.STRING, "auto")
+    val min_support = Parameter(AlgorithmParameterName.MIN_SUPPORT, DataType.DOUBLE, "0.5")
+    val min_confidence = Parameter(AlgorithmParameterName.MIN_CONFIDENCE, DataType.DOUBLE, "0.5")
 
     val parametersForAllAlgorithms = Seq(num_folds, max_parallelism, metric)
 
     Seq(
-      Algorithm(AlgorithmName.ARL_FPGROWTH, Some("FP Growth"), parametersForAllAlgorithms),
+      Algorithm(AlgorithmName.ARL_FPGROWTH, Some("FP Growth"), parametersForAllAlgorithms ++ Seq(min_support, min_confidence)),
       Algorithm(AlgorithmName.CLASSIFICATION_SVM, Some("Support Vector Machine (SVM)"), parametersForAllAlgorithms ++ Seq(max_iter, handle_invalid)),
       Algorithm(AlgorithmName.CLASSIFICATION_LOGISTIC_REGRESSION, Some("Logistic Regression"), parametersForAllAlgorithms ++ Seq(threshold, max_iter, reg_param, elasticnet_param, handle_invalid)),
       Algorithm(AlgorithmName.CLASSIFICATION_DECISION_TREE, Some("Decision Trees"), parametersForAllAlgorithms ++ Seq(max_depth, min_info_gain, impurity, handle_invalid)),

--- a/ppddm-manager/src/main/scala/ppddm/manager/controller/algorithm/AlgorithmController.scala
+++ b/ppddm-manager/src/main/scala/ppddm/manager/controller/algorithm/AlgorithmController.scala
@@ -18,6 +18,7 @@ object AlgorithmController {
     val num_folds = Parameter(AlgorithmParameterName.NUMBER_OF_FOLDS, DataType.INTEGER, "3")
     val max_parallelism = Parameter(AlgorithmParameterName.MAX_PARALLELISM, DataType.INTEGER, "2")
     val metric = Parameter(AlgorithmParameterName.METRIC, DataType.STRING, "areaUnderROC")
+    val handle_invalid = Parameter(AlgorithmParameterName.HANDLE_INVALID, DataType.STRING, "keep")
     val threshold = Parameter(AlgorithmParameterName.THRESHOLD, DataType.DOUBLE, "0.5")
     val max_iter = Parameter(AlgorithmParameterName.MAX_ITER, DataType.INTEGER, "10")
     val reg_param = Parameter(AlgorithmParameterName.REG_PARAM, DataType.DOUBLE, "0.0")
@@ -32,11 +33,11 @@ object AlgorithmController {
 
     Seq(
       Algorithm(AlgorithmName.ARL_FPGROWTH, Some("FP Growth"), parametersForAllAlgorithms),
-      Algorithm(AlgorithmName.CLASSIFICATION_SVM, Some("Support Vector Machine (SVM)"), parametersForAllAlgorithms ++ Seq(max_iter)),
-      Algorithm(AlgorithmName.CLASSIFICATION_LOGISTIC_REGRESSION, Some("Logistic Regression"), parametersForAllAlgorithms ++ Seq(threshold, max_iter, reg_param, elasticnet_param)),
-      Algorithm(AlgorithmName.CLASSIFICATION_DECISION_TREE, Some("Decision Trees"), parametersForAllAlgorithms ++ Seq(max_depth, min_info_gain, impurity)),
-      Algorithm(AlgorithmName.CLASSIFICATION_RANDOM_FOREST, Some("Random Forest"), parametersForAllAlgorithms ++ Seq(max_depth, min_info_gain, impurity, num_trees, feature_subset_strategy)),
-      Algorithm(AlgorithmName.CLASSIFICATION_GBT, Some("Gradient Boosted Trees"), parametersForAllAlgorithms ++ Seq(max_iter, max_depth, min_info_gain, feature_subset_strategy)),
+      Algorithm(AlgorithmName.CLASSIFICATION_SVM, Some("Support Vector Machine (SVM)"), parametersForAllAlgorithms ++ Seq(max_iter, handle_invalid)),
+      Algorithm(AlgorithmName.CLASSIFICATION_LOGISTIC_REGRESSION, Some("Logistic Regression"), parametersForAllAlgorithms ++ Seq(threshold, max_iter, reg_param, elasticnet_param, handle_invalid)),
+      Algorithm(AlgorithmName.CLASSIFICATION_DECISION_TREE, Some("Decision Trees"), parametersForAllAlgorithms ++ Seq(max_depth, min_info_gain, impurity, handle_invalid)),
+      Algorithm(AlgorithmName.CLASSIFICATION_RANDOM_FOREST, Some("Random Forest"), parametersForAllAlgorithms ++ Seq(max_depth, min_info_gain, impurity, num_trees, feature_subset_strategy, handle_invalid)),
+      Algorithm(AlgorithmName.CLASSIFICATION_GBT, Some("Gradient Boosted Trees"), parametersForAllAlgorithms ++ Seq(max_iter, max_depth, min_info_gain, feature_subset_strategy, handle_invalid)),
       // Algorithm(AlgorithmName.CLASSIFICATION_NAIVE_BAYES, Some("Naive Bayes"), parametersForAllAlgorithms), // Not working with existing BinaryClassificationEvaluator mechanism in Agents, hence removed from the list
       Algorithm(AlgorithmName.REGRESSION_LINEAR, Some("Linear Regression"), parametersForAllAlgorithms ++ Seq(threshold, max_iter, reg_param, elasticnet_param)),
       Algorithm(AlgorithmName.REGRESSION_DECISION_TREE, Some("Decision Trees"), parametersForAllAlgorithms ++ Seq(threshold, max_iter, reg_param, elasticnet_param)),

--- a/ppddm-manager/src/main/scala/ppddm/manager/controller/dm/AssociationMiningProcessor.scala
+++ b/ppddm-manager/src/main/scala/ppddm/manager/controller/dm/AssociationMiningProcessor.scala
@@ -139,8 +139,8 @@ object AssociationMiningProcessor {
 
         // Combine item frequencies and total record count
         val updatedBoostedModels = newDataMiningModel.boosted_models.get map { boostedModel =>
-          // Find the threshold value if it is provided in the GUI. Otherwise use the default value which is 0.3
-          var threshold = 0.3
+          // Find the threshold value if it is provided in the GUI. Otherwise use the default value which is 0.5
+          var threshold = 0.5
           val minSupport = boostedModel.algorithm.parameters.filter(p => p.name == AlgorithmParameterName.MIN_SUPPORT)
           if (!minSupport.isEmpty) {
             threshold = minSupport.head.value.toDouble

--- a/ppddm-manager/src/test/scala/ppddm/manager/api/endpoint/AlgorithmEndpointTest.scala
+++ b/ppddm-manager/src/test/scala/ppddm/manager/api/endpoint/AlgorithmEndpointTest.scala
@@ -21,7 +21,7 @@ class AlgorithmEndpointTest extends PPDDMManagerEndpointTest {
         status shouldEqual OK
 
         val response: Seq[Algorithm] = responseAs[Seq[Algorithm]]
-        response.size shouldEqual 11
+        response.size shouldEqual 10
         // We cannot make this check because we do not create the Enumarations appropriately. The values of the Enumeration objects are not Value instances
         // response.size shouldEqual AlgorithmName.values.size
       }

